File: dsa-problems/Trees/solutions/python/TRE-007-sports-dome-weighted-diameter.py
import sys

# Increase recursion depth
sys.setrecursionlimit(200000)

def weighted_diameter(n: int, left: list[int], right: list[int], lw: list[int], rw: list[int]) -> int:
    if n == 0:
        return 0
        
    max_diameter = 0
    
    def dfs(u):
        nonlocal max_diameter
        if u == -1:
            return 0
            
        l_path = 0
        r_path = 0
        
        if left[u] != -1:
            l_path = lw[u] + dfs(left[u])
        if right[u] != -1:
            r_path = rw[u] + dfs(right[u])
            
        max_diameter = max(max_diameter, l_path + r_path)
        
        return max(l_path, r_path)
        
    # Find root - node with no incoming edges (no parent)
    has_parent = [False] * n
    for i in range(n):
        if left[i] != -1: has_parent[left[i]] = True
        if right[i] != -1: has_parent[right[i]] = True
        
    root = 0
    for i in range(n):
        if not has_parent[i]:
            root = i
            break
            
    dfs(root)
    return max_diameter

def main():
    lines = sys.stdin.read().splitlines()
    if not lines: return
    valid_lines = [l.strip() for l in lines if l.strip()]
    if not valid_lines: return
    iterator = iter(valid_lines)
    
    try:
        n = int(next(iterator))
        left = [0] * n
        right = [0] * n
        lw = [0] * n
        rw = [0] * n
        
        for i in range(n):
            line = next(iterator)
            parts = list(map(int, line.split()))
            left[i] = parts[1]
            right[i] = parts[2]
            if len(parts) >= 5:
                lw[i] = parts[3]
                rw[i] = parts[4]
            else:
                lw[i] = 1
                rw[i] = 1
                
        print(weighted_diameter(n, left, right, lw, rw))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/Trees/solutions/javascript/TRE-007-sports-dome-weighted-diameter.js
const fs = require("fs");

const lines = fs
  .readFileSync(0, "utf8")
  .split(/\r?\n/)
  .map((line) => line.trim())
  .filter((line) => line.length > 0);

if (lines.length === 0) {
  process.exit(0);
}

const n = parseInt(lines[0], 10);
if (n === 0) {
  console.log("0");
  process.exit(0);
}

const left = new Array(n).fill(-1);
const right = new Array(n).fill(-1);
const lw = new Array(n).fill(1);
const rw = new Array(n).fill(1);

for (let i = 0; i < n && i + 1 < lines.length; i++) {
  const parts = lines[i + 1].split(/\s+/).map(Number);
  if (parts.length < 3) continue;
  left[i] = parts[1];
  right[i] = parts[2];
  if (parts.length >= 5) {
    lw[i] = parts[3];
    rw[i] = parts[4];
  }
}

const hasParent = new Array(n).fill(false);
for (let i = 0; i < n; i++) {
  if (left[i] !== -1) hasParent[left[i]] = true;
  if (right[i] !== -1) hasParent[right[i]] = true;
}
let root = 0;
for (let i = 0; i < n; i++) {
  if (!hasParent[i]) {
    root = i;
    break;
  }
}

const stack = [root];
const order = [];
const visited = new Array(n).fill(false);
visited[root] = true;

while (stack.length > 0) {
  const u = stack.pop();
  order.push(u);
  const l = left[u];
  const r = right[u];
  if (l !== -1 && !visited[l]) {
    visited[l] = true;
    stack.push(l);
  }
  if (r !== -1 && !visited[r]) {
    visited[r] = true;
    stack.push(r);
  }
}

const dist = new Array(n).fill(0);
let maxDiameter = 0;
for (let i = order.length - 1; i >= 0; i--) {
  const u = order[i];
  let lPath = 0;
  let rPath = 0;
  if (left[u] !== -1) lPath = lw[u] + dist[left[u]];
  if (right[u] !== -1) rPath = rw[u] + dist[right[u]];
  const dia = lPath + rPath;
  if (dia > maxDiameter) maxDiameter = dia;
  dist[u] = lPath > rPath ? lPath : rPath;
}

console.log(maxDiameter.toString());
